### ISSUE TITLE
Investigate compressions

### ISSUE THREAD

**Issue by puzrin** [littlevgl/lv_font_conv#8]
Need to investigate possible benefits of compression use.

## Criterias

- Base level to compare: uncompressed data, but with cropped bitmaps.
- "Settings menu" with Material Design theme should fit into 128K stm32f0
- Smoothness requirement: 2-3 bpp
- Don't care about small 1 bpp font size for bi-tonal OLED-s.

## Algorythms

- [I3BN](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwi35uLL_LLhAhVCK1AKHQeLDXEQFjAAegQIAxAB&url=https%3A%2F%2Fwww.researchgate.net%2Fpublication%2F305820775_New_Modified_RLE_Algorithms_to_Compress_Grayscale_Images_with_Lossy_and_Lossless_Compression&usg=AOvVaw0ZI6dNRXRsGUx6wPN0lBbj) (RLE), restricted to "white" or "white" + "black" only. Should be very simple and effective
  - For low bpp - additionally restrict with minimal required repeat count
- Filtering
  - Lines delta coding. May help in some cases. For effective implementation - align output lines to word, and apply binary XOR
  - BWT (just for fun). Decoder requires additional memory to sort data.

This combinations should give optimal balance between compression ratio and hardware requirements.

**Comment by puzrin** [comment]
Correction:

- Optimal bpp is 3 (2 - not enougth)

**Comment by puzrin** [comment]
**Ideas**

## Huffman encoding

Could we apply Huffman encoding somehow? Probably, with predefined dictionary. Table-based decoder is very cheap.

- In theory, this may replace I3BN if long repetitions are encoded as "super-pixel"
- At first glance, should not help too much after I3BN.

## Word-aligned output lines

- If output buffer has word-aligned lines, it will be easy to apply delta-encoding to lines. Since we care about exact vertical repetitions only, binary `XOR` will be best solution.

**Comment by puzrin** [comment]
Added `--dump-delta` to apply `XOR` filter to lines. Result looks promissing.

## Compression

After analyzing bitmaps and delta-line bitmaps, "invented" better condition to restrict RLE overhead.

Instead of apply RLE selectively to Black and White only, apply it everywhere, BUT request minimal repetitions count to enable:

- [ a, a, b ] => [ a, a, b ]
- [ a, a, a, a, b ] => [ a, a, rle_comressed, b ]

In other words, short repetitions will not be compressed. That's ok, because:

- halftones have no short repetitions at all (and rare long repetitions)
- whites & blacks have longer repetitions

**Comment by kisvegabor** [comment]
Huffman encoding with XOR sound very promising. I'm really looking forward to see the compression rate!

> Instead of apply RLE selectively to Black and White only, apply it everywhere, BUT request minimal repetitions count to enable:
[ a, a, b ] => [ a, a, b ]
[ a, a, a, a, b ] => [ a, a, rle_comressed, b ]

Why not `[ a, a, a, a, b ] => [ rle_comressed, b ]` ?

**Comment by puzrin** [comment]
> Why not `[ a, a, a, a, b ] => [ rle_comressed, b ]` ?

Because there are lot of "single" (non-continuous) pixels, with 3 bits. If you add "stop" bit after each, overhead will be significant. With new restriction overhead for "random" sequences will be zero.

**Comment by kisvegabor** [comment]
I see, thank you.

**Comment by puzrin** [comment]
`bin` branch contains stub of but writer. Not complete, bit tested and without compression. But enougth to investigate memory distribution.

**16px, 0x20-0x7f, 3bpp**

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 16 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -4 +3ms
  font maxY: 14 +0ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +1ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 9 +0ms
  font indexToLocFormat: 0 +5ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 3456 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +1ms
  font font size: 5136 +12ms
```

**24px, 0x20-0x7f, 3bpp**

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 24 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -6 +2ms
  font maxY: 20 +0ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +0ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 10 +1ms
  font indexToLocFormat: 0 +10ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 6940 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +0ms
  font font size: 8620 +12ms
```

Preliminary conclusion:

- Kerning size is significant (and 4x more, 5K, for 0x20-0xFE range). We can not expect significant reduce for 16px fonts, but still can have benefits with bugger fonts.
- Only "high quality fonts" are considered above. Size of bitonal monospace oled fonts is already small enougth.
- Material design seems realistic with 128k flash SOC-s.

I will post update when add compression.

@kisvegabor could you tell aproximate size of lvgl5 fonts? (range 0x20-0x7f, bpp4, sizes 16px and 24px)

**Comment by puzrin** [comment]
Added option `--no-compression`, to compare savings. Results below are with enabled modified RLE.

**16px, 0x20-0x7f, 3bpp**

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 16 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -4 +3ms
  font maxY: 14 +0ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +2ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +0ms
  font wh_bits: 5 +1ms
  font advanceWidthBits: 9 +0ms
  font indexToLocFormat: 0 +17ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 2956 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +1ms
  font font size: 4636 +10ms
```

without lines xor:

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 16 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -4 +2ms
  font maxY: 14 +0ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +0ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 9 +1ms
  font indexToLocFormat: 0 +8ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 3248 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +0ms
  font font size: 4928 +13ms
```

**24px, 0x20-0x7f, 3bpp**

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 24 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -6 +2ms
  font maxY: 20 +1ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +0ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 10 +0ms
  font indexToLocFormat: 0 +20ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 5200 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +1ms
  font font size: 6880 +10ms
```

without lines xor:

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 24 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -6 +2ms
  font maxY: 20 +0ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +1ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 10 +0ms
  font indexToLocFormat: 0 +11ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 5904 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +0ms
  font font size: 7584 +15ms
```

---

Conclusion:

- Notable difference can be seen on large sizes only (30-40px).
- Is we use bit streams to not align each line, dropping RLE machinery will not save much of code.

Note, this compare does not count savings from cropping bitmaps by bounding box.

**Comment by puzrin** [comment]
More generic compare. Input data already cropped by bbox.

24px, 0x20-0x7f:

- bpp 4, no-compression => 10792 bytes
- bpp 4, compressed => 8308 bytes
- bpp 3, no-compression => 8620 bytes
- bpp 3, compressed => 6880 bytes

16px, 0x20-0x7f:

- bpp 4, no-compression => 6162 bytes
- bpp 4, compressed => 5636 bytes
- bpp 3, no-compression => 5136 bytes
- bpp 3, compressed => 4636 bytes

**Comment by puzrin** [comment]
Tuned compression settings for our kind of data:

1. Repetitions without counter write: 2 => 10
2. Counter bits: 6 (best)

It still works for `--bpp 2` mode (result is NOT worse), but anyway, this mode is not very useful. Optimal are `--bpp 1` for OLEDs & `--bpp 3` for TFT/IPS.

For `--bpp 1` compression is disabled. It needs different approach. Also, IMHO savings will be small and not worth time been spend.

**Comment by kisvegabor** [comment]
DeJavu 16px, 4 bpp: 4912 byte
DeJavu 24px, 4 bpp: 11184 bytes

The size of the 16 px font is very small compared to yours. If the 24 px font is 1.5 times wider and taller its size should be 4912 x 1.5 x 1.5 = 11,052 bytes. So these numbers seem correct. Why is your 16 px font so large event with bounding box and compression?

**Comment by puzrin** [comment]
> Why is your 16 px font so large event with bounding box and compression?

Let's clarify what you mean as "so large":

```
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --bpp 3 --size 16 --format bin -o test.bin
  font last_id: 96 +0ms
  font minY: -4 +2ms
  font maxY: 14 +1ms
  font glyphIdFofmat: 0 +0ms
  font kerningFormat: 0 +1ms
  font advanceWidthFormat: 1 +0ms
  font xy_bits: 5 +0ms
  font wh_bits: 5 +0ms
  font advanceWidthBits: 9 +0ms
  font monospaced: false +0ms
  font indexToLocFormat: 0 +9ms
  font.table.head table size = 36 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 2836 +0ms
  font.table.kern 434 kerning pairs used +0ms
  font.table.kern table size = 1316 +0ms
  font font size: 4516 +15ms
```

1. bpp 3, compressed => **4516** bytes. That **includes 1.3K kerning** table. Actual, data size is smaller. If you drop kerning, difference will be significant.
2. Visual size (in pixels) after old convertor is smaller than required. If you order "16px", actual size should overflow this limits. So, new font has bigger bitmaps and require more bits to encode. And still smaller :)

So, i can say:

- For 16px you will not get size saving but will get quality increase (kerning).
- For 24px, new 6628 bytes against old 11184 bytes seems significant difference.
- With IPS 240x240 (high DPI), MD will require 30-40px fonts, and difference will be huge.

Anything not ok?

**Comment by kisvegabor** [comment]
I see, makes sense now!

**Comment by puzrin** [comment]
> Finally, in v6.0 format the exporting of the kerning table should be optional to make possible to save memory instead of better quality. I think it's not a problem.

That's doable, but i strongly not recommend to go this way. IMHO proper solution would be to use monospace fonts instead of such hacks. They have no or small kerning.

**Comment by embeddedt** [comment]
@puzrin But are there Sans/Sans-Serif fonts available that are monospaced? Most monospaced fonts look lower-quality in general because of their intended use case.

**Comment by puzrin** [comment]
@embeddedt I'm sure, yes. Example - Terminus + something from X11.

I suggest to view on this problem from different language. Users don't care about bytes, they wish "good looking interface". So, choice are:

1. Take old-school pixel-perfect font (X11 and so on). Those are not scaleable and are only for fixed size. Probably, those available in BDF format (not supported now, but can add later).
2. Take modern monospace TTF font (such fonts are widely used in code editors).
3. Take modern proportional font, with kerning.

If you take proportional TTF and drop kerning - result will be bad with guarantee. From my point of view - it's a dirty hack, that should not exist :)

**Comment by embeddedt** [comment]
> If you take proportional TTF and drop kerning - result will be bad with guarantee.

Is that what the v5.x convertor is doing now?

**Comment by puzrin** [comment]
> Is that what the v5.x convertor is doing now?

Yes. And that breaks my eyes :)

**Comment by embeddedt** [comment]
@puzrin Hmm, I'm still not sure we should require the kerning table. No one has complained thus far about the quality of the text. A few people have asked for kerning, but overall everyone seems happy with it.

Should we really be forcing them to use additional memory for the kerning information?

**Comment by puzrin** [comment]
> Should we really be forcing them to use additional memory for the kerning information?

At first, new format will take less space, even with kerning table.

At second, from my point of view, many things in "embedded" looks horrible :). Just because everything is done by single person, and one person can not be good programmer and good designer in the same time.

I guess, "no complains" is not because everything is perfect, but because people don't know how good it can be. But after you see good result once, you will never wish to return back. There are myths, that small SOC-s have few resources and being ugly is normal trade-off. I don't agree and try to fix most critical parts:

- i18n - done
- fonts - in progress
- settings menu - in queue (subset of Material Design, doable in restricted time).

**Comment by puzrin** [comment]
In short - font designers do the best to make font look properly. Programmers should not try to "improve" fonts programmatically. If somebody wish font without kerning - take appropriate source font, and be happy :)

**Comment by kisvegabor** [comment]
Actually, I agree with both of you. 

There are so much use cases and there is no reason to make kerning table compulsory.  

However, I'm sure it will make texts drawn by LittelvGL much better. Somewhat closer to non-embedded GUIs. 

Making possible to enable/disable kerning would be great. This way people could easily see the difference and decide if it is worth them or not.

**Comment by puzrin** [comment]
As i said, if you tell me "i need this feature" - i will add this. My point is, that focusing on "easy way" will prevent us from search way to "do things right".

https://github.com/littlevgl/lvgl/issues/1004 - this issue was created exactly to help people find right kerning-less monochrome sources, instead of try use "modern" TTF fonts how those are not expected to be used (without kerning)

**Comment by embeddedt** [comment]
@puzrin We can still "do things the right way" by supporting the kerning table, but not requiring it.

@kisvegabor Where did the builtin `Dejavu` fonts originate from? Were they created from TTF files?

**Comment by puzrin** [comment]
@embeddedt 

> We can still "do things the right way" by supporting the kerning table, but not requiring it.

You mix different things. There are 2 kind of fonts: with kerning and without kerning. Those are designed differently. My suggestion is to select proper source, instead of hacking TTF. That's a right way from my point of view.

**Comment by embeddedt** [comment]
@puzrin

> My suggestion is to select proper source, instead of hacking TTF. That's a right way from my point of view.

Yes, but if users are happy with the slightly incorrect way (TTF fonts with kerning ignored), should we force them to do things the "right way"? I'm not saying kerning shouldn't be supported; I'm saying it shouldn't be a requirement.

Another example: we dropped ASCII support, but this frustrated a user because it broke support for 1-byte character encodings. It's a similar situation. Even though only supporting UTF-8 would be a cleaner and simpler design, if it's not what the users want, we should prioritize user needs over "clean design", to a point.

I think the same principle applies to both cases. While cleaning up the font design is a good idea, we shouldn't do it in a way that gives the users extra burden.

**Comment by puzrin** [comment]
Such "features" has historic reasons - when "proper" renderer was not available, people tried have "something working". Not because this is correct, and not because people have no memory - just because they did not know how to do right (and that's not their fault).

It total, that cause compatibility issues when we try to do things right. See questions about hard-pinned functions into font, hook for "glyph rendering via frrtype" and so on. We just managed to cleanup old mess, and i'd like to not introduce new one :). If you don't promote kerning drop, and explain where to take pixel-perfect fonts - people will be happy, IMO.

**Comment by kisvegabor** [comment]
In summary:
- "with kerning" should be the default  option even if it's still new people
- kerning will be hidden from the users, they just get better looking texts
- the kerning table has a memory overhead. I was ~1.3kB for the ASCII range with Robot font
- there are too much use cases and we can't think about everything, therefore, we should provide the highest flexibility where possible
- to work without kerning table is simple both in LittlevGL and the Font converter so it's worth to make it optional (can be enabled by default)

Do you agree with it?

**Comment by puzrin** [comment]
https://github.com/littlevgl/lv_font_conv/commit/2a915e09a9fcc3a7f102db171c93aaf7ef81aa8e

As i said, i prefer to search solutions without quality loss. But it costs me nothing to add this option.

One of possible saving could be to invent "font pack" with shared tables. For example, if you you need fonts of multiple sizes (16px, 24px, 32px) - only `glyph` table must be unique, others can be shared and rescaled on the fly.

**Comment by kisvegabor** [comment]
> One of possible saving could be to invent "font pack" with shared tables. For example, if you you need fonts of multiple sizes (16px, 24px, 32px) - only glyph table must be unique, others can be shared and rescaled on the fly.

Sounds interesting, but let's postpone it for a next release. :)

**Comment by puzrin** [comment]
Sure. I'm satisfied with current font format. Let's wait for new hero to pass this quest again :)

Close as solved.
